File: golang/day1/main.go
package main

import (
	"fmt"
	"math"
	"strings"
	"time"
)

// Defining a struct
type Person struct {
	Name string
	Age  int
}

// Method for the struct
func (p Person) Introduce() {
	fmt.Printf("Hi, I'm %s and I'm %d years old.\n", p.Name, p.Age)
}

// Defining an interface
type Shape interface {
	Area() float64
}

// Struct implementing the Shape interface
type Circle struct {
	Radius float64
}

func (c Circle) Area() float64 {
	return math.Pi * c.Radius * c.Radius
}

// Function with multiple return values
func divideAndRemainder(a, b int) (int, int) {
	return a / b, a % b
}

func main() {
	// Printing to console
	fmt.Println("Hello, world!")

	// Variables and short declaration
	var x int = 5
	y := 10

	// Basic types
	var (
		integer int     = 42
		float   float64 = 3.14
		boolean bool    = true
		char    rune    = 'A'
	)
	fmt.Println(
		integer,
		float,
		boolean,
		char,
	)

	// Strings
	str1 := "Hello"
	str2 := "World"
	fmt.Println(strings.Join([]string{str1, str2}, " "))

	// Slices
	numbers := []int{1, 2, 3, 4, 5}
	fmt.Println("Slice:", numbers)

	// Maps
	ages := map[string]int{
		"Alice": 30,
		"Bob":   25,
	}
	fmt.Println("Bob's age:", ages["Bob"])

	// Control flow
	if x < 10 {
		fmt.Println("x is less than 10")
	} else {
		fmt.Println("x is 10 or greater")
	}

	// Switch statement
	switch y {
	case 5:
		fmt.Println("y is 5")
	case 10:
		fmt.Println("y is 10")
	default:
		fmt.Println("y is neither 5 nor 10")
	}

	// For loop
	for i := 0; i < 5; i++ {
		fmt.Printf("Loop iteration: %d\n", i)
	}

	// Range loop
	for index, value := range numbers {
		fmt.Printf("Index: %d, Value: %d\n", index, value)
	}

	// Function call with multiple return values
	quotient, remainder := divideAndRemainder(17, 5)
	fmt.Printf("17 divided by 5 is %d with remainder %d\n", quotient, remainder)

	// Using the struct
	person := Person{Name: "Alice", Age: 30}
	person.Introduce()

	// Using the interface
	circle := Circle{Radius: 5}
	fmt.Printf("Circle area: %.2f\n", circle.Area())

	// Error handling
	_, err := time.Parse("2006-01-02", "2023-13-01")
	if err != nil {
		fmt.Println("Error parsing date:", err)
	}

	// Defer statement
	defer fmt.Println("This will be printed last")

	// Goroutine and channel
	ch := make(chan string)
	go func() {
		ch <- "Hello from goroutine!"
	}()
	fmt.Println(<-ch)

	fmt.Println("Main function completed")
}

File: rust/day1/src/main.rs
// Importing standard library modules
use std::collections::HashMap;

// Defining a struct
struct Person {
    name: String,
    age: u32,
}

// Implementing methods for the struct
impl Person {
    fn new(name: String, age: u32) -> Person {
        Person { name, age }
    }

    fn introduce(&self) {
        println!("Hi, I'm {} and I'm {} years old.", self.name, self.age);
    }
}

// Defining an enum
enum Color {
    Red,
    Green,
    Blue,
}

// Main function
fn main() {
    // Printing to console
    println!("Hello, world!");
    println!("Hello, {}!", "world");
    print!("Hello, world!\n");

    let name = "John";
    let game = "football";

    println!("{name} like to play {game}", name = name, game = game);
    println!("{} like to play {}", name, game);
    println!("Binary: {:b} - Hex: {:x} - Octal {:o}", 10, 10, 10);

    // Variables and mutability
    let immutable_var1 = 5;
    let mut mutable_var1 = 10;
    mutable_var1 += 5;

    let mut mutable_float1 = 10.;
    mutable_float1 += 3.;

    println!("mutable_float1 = {}", mutable_float1);

    // Basic types
    let integer: i32 = 42;
    let float: f64 = 3.14;
    let boolean: bool = true;
    let character: char = 'A';

    // Strings
    let string_literal = "Hello";
    let string = String::from("World");

    // Arrays and Vectors
    let array = [1, 2, 3, 4, 5];
    let vector: Vec<i32> = vec![1, 2, 3, 4, 5];

    // Control flow
    if immutable_var1 < 10 {
        println!("Less than 10");
    } else {
        println!("10 or greater");
    }

    // Loops
    for i in 0..5 {
        println!("Loop iteration: {}", i);
    }

    let mut counter = 0;
    while counter < 3 {
        println!("While loop: {}", counter);
        counter += 1;
    }

    // Pattern matching
    let color = Color::Red;
    match color {
        Color::Red => println!("It's red!"),
        Color::Green => println!("It's green!"),
        Color::Blue => println!("It's blue!"),
    }

    // Functions
    fn add(a: i32, b: i32) -> i32 {
        a + b
    }
    let sum = add(5, 7);
    println!("Sum: {}", sum);

    // Closures
    let multiply = |x: i32, y: i32| x * y;
    println!("Product: {}", multiply(3, 4));

    // Using the struct
    let person = Person::new(String::from("Alice"), 30);
    person.introduce();

    // Error handling
    let result: Result<i32, &str> = Ok(42);
    match result {
        Ok(value) => println!("Success: {}", value),
        Err(error) => println!("Error: {}", error),
    }

    // Option type
    let maybe_value: Option<i32> = Some(42);

    if let Some(value) = maybe_value {
        println!("Got a value: {}", value);
    }

    // HashMap
    let mut scores = HashMap::new();
    scores.insert(String::from("Blue"), 10);
    scores.insert(String::from("Red"), 50);

    // Ownership and borrowing
    let owned_string = String::from("Hello");
    let borrowed_string = &owned_string;
    println!("Borrowed: {}", borrowed_string);

    // Slices
    let slice = &array[1..4];
    println!("Slice: {:?}", slice);
}
